### PR TITLE
Allow logging outside of setting the logging callback

### PR DESCRIPTION
**Pull request by deschuma**
- Addresses Issue #102 to enable logging to stdout independent of OE SDK
- Implements OE SDK ask to enable testing without on-demand registration, if desired
- Moves the environment variable functionality into a header file

### CODE REVIEW

File: src/environment.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            (strnlen(env_value, MAX_ENV_VAR_LENGTH) == MAX_ENV_VAR_LENGTH))
+        {
+            log(SGX_QL_LOG_ERROR,
+                "Value specified in environment variable %s is either empty or "
+                "expected max length '%d'.",
+                env_variable.c_str());
+            return std::string();
+        }
+
+        auto retval = std::string(env_value);
```
Comment (LarryOsterman) [inline src/environment.h:62]: Why save to a local variable here before returning? Why not just `return std::string(env_value);`?

File: src/environment.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        env_variable.c_str(), env_temp.get(), MAX_ENV_VAR_LENGTH);
+    if (status == 0)
+    {
+        log(SGX_QL_LOG_ERROR,
+            "Failed to retreive environment varible for '%s'",
+            env_variable.c_str());
+        return std::string();
+    }
+#endif
+    else
```
Comment (yentsanglee) [inline src/environment.h:50]: >else [](start = 4, length = 4)

This 'else' is unnecessary.  Both 'if' statements exit the function. Visually, it seems odd to share 'else' block by 2 'if' blocks.

File: src/environment.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        if ((strnlen(env_value, MAX_ENV_VAR_LENGTH) <= 0) ||
+            (strnlen(env_value, MAX_ENV_VAR_LENGTH) == MAX_ENV_VAR_LENGTH))
+        {
+            log(SGX_QL_LOG_ERROR,
+                "Value specified in environment variable %s is either empty or "
+                "expected max length '%d'.",
```
Comment (yentsanglee) [inline src/environment.h:57]: >%d [](start = 38, length = 2)

Missing an integer param for the %d.

**Comment by soccerGB** [comment]
The logging for parameters passed in sgx_ql_get_quote_config() is only on when the target was Not found in the cache. Could you take this chance to enable logging when cache was hit as well?

File: src/logging.cpp [hunk 4]
```diff
@@ -15,36 +15,117 @@
...
 #include <vector>
+#include <mutex>
+#include "environment.h"
 
-#if 0 // Flip this to true for easy local debugging
-static void DefaultLogCallback(sgx_ql_log_level_t level, const char* message)
+using namespace std;
+
+sgx_ql_logging_function_t logger_callback = nullptr;
+static sgx_ql_log_level_t debug_log_level;
```
Comment (yentsanglee) [inline src/logging.cpp:24]: >debug_log_level [](start = 26, length = 15)

Initialize it.

File: src/logging.cpp [hunk 5]
```diff
@@ -15,36 +15,117 @@
...
+// Logging function which doesn't allocate any memory
+//
+void log_message(sgx_ql_log_level_t level, const char* message)
+{
+    init_debug_log();
 
     if (logger_callback != nullptr)
     {
         logger_callback(level, message);
+    }else if (enable_debug_log)
```
Comment (yentsanglee) [inline src/logging.cpp:123]: Linefeed

File: src/logging.cpp [hunk 6]
```diff
@@ -15,36 +15,117 @@
...
+#include <mutex>
+#include "environment.h"
 
-#if 0 // Flip this to true for easy local debugging
-static void DefaultLogCallback(sgx_ql_log_level_t level, const char* message)
+using namespace std;
+
+sgx_ql_logging_function_t logger_callback = nullptr;
+static sgx_ql_log_level_t debug_log_level;
+bool enable_debug_log = false;
```
Comment (yentsanglee) [inline src/logging.cpp:25]: >enable_debug_log [](start = 5, length = 16)

Minor suggestion - enable_debug_log is always coupled with debug_log_level. You can define something like "SGX_QL_LOG_NONE = 0" to replace enable_debug_log being false.

File: src/environment.h (hunk 3 above)
Comment (deschuma) [reply src/environment.h:57]: Thanks fixed

File: src/environment.h (hunk 1 above)
Comment (deschuma) [reply src/environment.h]: Thanks fixed

File: src/logging.cpp (hunk 6 above)
Comment (deschuma) [reply src/logging.cpp]: thanks - fixed

File: src/logging.cpp (hunk 5 above)
Comment (deschuma) [reply src/logging.cpp]: thanks - fixed

**Comment by deschuma** [comment]
@soccerGB logging for parameters happens regardless of caching or not - is there a specific parameter I am missing?
1: Azure Quote Provider: libdcap_quoteprov.so [INFO]: Fetching quote config from cache: 'https://global.acccache.azure.net/sgx/certificates/v2/e40610aa353e1bb52385aba37e8acadb/0f0f0205ff8007000000000000000000/0a00/0000?clientid=production_client&api-version=2018-10-01-preview'.

**Pull request merged**